### ISSUE TITLE
Class static block does not allow forward reference to static method

### ISSUE THREAD

**Issue by fwienber** [microsoft/TypeScript#45932]
# Bug Report

### 🔎 Search Terms

class static blocks, forward reference, incorrect compiler error

### 🕗 Version & Regression Information

4.4.2 + nightly (4.5.0-dev20210910)

- I was unable to test this on prior versions because the feature is new in 4.4

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.4.2#code/MYGwhgzhAECC0G8CwAoa7oQC5iwS2EVQxLgDoATAewGUqBbAUywAs8A7AcwAoBKAbmgB6IdACiAJwlUJALmgAiAArSADowlYAntADk1Ok1YdOu6HhgBXCIwrQARowBmMxuawwOefGBB4AXrh4VOxk3ABMAOzhAJy8CsQYAL6oiejYQYQGDMxsXHxEaKTQwCEQVCCMZCBUPAqcVFjALGAAhAoCadApKElAA)

### 💻 Code

```ts
class A {
    static {
        A.doSomething();
    }

    static doSomething() {
       console.log("gotcha!");
    }
}
```

### 🙁 Actual behavior

Compiler error in line 3: "Property 'doSomething' is used before its initialization.(2729)"

The code works as-is in native  JavaScript, tested in Firefox Developer Edition.

The rewrites for class static blocks in Babel as well as in TypeScript itself also work. Both execute the static block's code at a time where static methods of the class are already available.

### 🙂 Expected behavior

No compiler error, because everything's fine.

A forward reference to a static _field_, however, should be an error (I guess).

### Workaround

Fall back to the way class static blocks could be emulated before they were supported:
Define an unused #-private static constant whose initializer is an immediate-evaluating function.

```ts
class A {
    static #_ = (() => {
        A.doSomething();
    })();

    static doSomething() {
       console.log("gotcha!");
    }
}
```

[Playground link for workaround](https://www.typescriptlang.org/play?ts=4.4.2#code/MYGwhgzhAECC0G8CwAoa7oQC5iwS2GgGIB9aAXmgAoqBKCgPkVQ1bgDoATAewGVuAtgFMsACzwA7AOZ0A3CwwBfWnNQL02XAWg9+wsZJn1kaNtGDcJEbiCHsQ3GQCIp3LMFFgAhE9rzT6IqoikA)

Looks ugly, but works in the desired execution order and produces no compiler error.

### Possibly related issues

- Execution order of static field initialization and static initialization blocks is not kept under target: esnext and useDefineForClassFields: false #45574
- Statements in class static blocks do not narrow types #44949

**Comment by fwienber** [comment]
Any first evaluation / labels for this one? Maybe @rbuckton or @Zzzen?
I hope the example code makes the case quite obvious.

**Comment by fwienber** [comment]
Thank you for adding this to 4.6.0 milestone.
I was just wondering why this still has the "needs investigation" label. Do you doubt this is a bug? Do you need further elaboration?

**Comment by andrewbranch** [comment]
Issues are currently coming in faster than I can triage them. It looked like a probable bug to me but I know very little about class static blocks. If I were to take the time to be 100% confident that this is a bug, I would not have time to do the things that need to happen to release TypeScript 4.5. On the other hand, Ron will know if this is a bug within about 30 seconds. “Needs investigation” ~= “Looks suspicious but I defer to the assignee’s judgment, because they have more expertise than me, can devote more time to studying it than me, or both.” Sadly that doesn’t fit in a label.

**Comment by rbuckton** [comment]
Methods should be accessible. Static fields and static blocks are deferred until the end of ClassDefinitionEvaluation after static and prototype methods are defined (including private static elements).

**Comment by fwienber** [comment]
To motivate that this should _really_ be included in the next milestone: It's even worse!
I just found out that from a static code block, you cannot even reference a constant defined _before_ the class:
````TypeScript
const FOO = "FOO";
class Bar {
    static {
        console.log(FOO); // Variable 'FOO' is used before being assigned.(2454)
    }
}
````
[Run in TypeScript Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAYgeQTAvDARIh6DcAoYAGwEMIIYAhYgJxgG88YmZpioBLYex530SEIQCmAOkIgA5gAosASny8AvnmVA)

This results in the unjustified compile error "Variable 'FOO' is used before being assigned.(2454)", but if you still run it, works as it should, logging "FOO".
For this additional case, I can't even think of a work-around, which makes this bug far more serious.

**Comment by Zzzen** [comment]
You are probably the first person in the world who use class static block with TypeScript.

**Comment by Zzzen** [comment]
> To motivate that this should really be included in the next milestone: It's even worse!
> I just found out that from a static code block, you cannot even reference a constant defined before the class:
> 
> ```ts
> const FOO = "FOO";
> class Bar {
>     static {
>         console.log(FOO); // Variable 'FOO' is used before being assigned.(2454)
>     }
> }
> ```
> 
> [Run in TypeScript Playground](https://www.typescriptlang.org/play?#code/MYewdgzgLgBAYgeQTAvDARIh6DcAoYAGwEMIIYAhYgJxgG88YmZpioBLYex530SEIQCmAOkIgA5gAosASny8AvnmVA)
> 
> This results in the unjustified compile error "Variable 'FOO' is used before being assigned.(2454)", but if you still run it, works as it should, logging "FOO".
> For this additional case, I can't even think of a work-around, which makes this bug far more serious.

This is related to #44949. so, you are probably the second one.
